### DISCUSSION TITLE
rk3576 development

### DISCUSSION THREAD

**Discussion by dsowa** [Joshua-Riek/ubuntu-rockchip#959] [General]
I see from the commits that the Sige5 board has been thought about at least a little bit. I'm definitely interested in the RK3576 soc and since the Sige5 seems to be the only current SBC out there, support for that board as well.  Are there any thoughts on where the development for this is heading?  Are there holdups on the Rockchip side? Thanks.

**Comment by Joshua-Riek** [comment]
I really like the board design, I only know of two issues the RK3576 SoC has right now.

- The GPU driver Panfrost does not work.
- SD Cards do not work in Linux (I was able to fix this in U-Boot, but I need to investigate further for Linux kernel support).

I'm shocked that Panfrost does not work as Mali-G52 has been supported for a while. Collabora says the GPU is supported since Linux 5.10-rc1, but a device tree binding is needed for mainline, which leaves me confused as the Rockchip SDK is 6.1, so in theory it should work. 

I would love to dig into this silly GPU situation, but my GPU driver knowledge is not very good.

**Reply by nyanmisaka** [reply]
First check if panfrost KMD is loaded correctly, otherwise it might have some kind of hardware quirk.
- https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/panfrost/panfrost_drv.c#L726-L752
- https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/panfrost/panfrost_gpu.c#L104-L113

Then check if it is a new variant of G52 that Mesa does not recognize yet.
- https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/panfrost/lib/pan_props.c#L66-L67
- https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10771

**Reply by Joshua-Riek** [reply]
My first thought was wondering if mesa does not detect the new variant, as we can use gles2 with the mali blob. Thanks for the tips. I'll check them out!

**Comment by Joshua-Riek** [comment]
Ah, may be more complicated, it looks like the RK3576 GPU is actually deferred. Need to figure out what the problem is with the DCDC_REG5 regulator. 
```
ubuntu@ubuntu:~$ sudo cat /sys/kernel/debug/devices_deferred 
hdmi-sound	
mtd_vendor_storage	
vcc-1v8-s0	platform: wait for supplier DCDC_REG8
vcc-3v3-s0	platform: wait for supplier DCDC_REG4
vcc1v8-ufs-vccq2-s0	platform: wait for supplier DCDC_REG8
27800000.gpu	platform: wait for supplier DCDC_REG5
dmc	platform: wait for supplier DCDC_REG7
2ae00000.adc	platform: wait for supplier PLDO_REG1
es8388-sound	asoc-simple-card: parse error
27700000.npu	platform: wait for supplier DCDC_REG2
2a310000.mmc	platform: wait for supplier PLDO_REG5
ubuntu@ubuntu:~$ cat /sys/devices/platform/27800000.gpu/waiting_for_supplier 
1
```

**Reply by Joshua-Riek** [reply]
Seems to be a problem with the Ubuntu Rockchip kernel. I have a test below using the factory Debian 12 image and we can see the GPU loading correctly. 
```
armsom@armsom-sige5:~$ dmesg | grep -i gpu
[    4.183992] mali 27800000.gpu: Kernel DDK version g21p0-01eac0
[    4.184211] mali 27800000.gpu: GPU metrics tracepoint support enabled
[    4.184605] mali 27800000.gpu: bin=0
[    4.184811] mali 27800000.gpu: leakage=5
[    4.184835] debugfs: Directory '27800000.gpu-mali' with parent 'vdd_gpu_s0' already present!
[    4.186521] mali 27800000.gpu: pvtm=816
[    4.187028] mali 27800000.gpu: pvtm-volt-sel=1
[    4.187069] debugfs: Directory '27800000.gpu-mali' with parent 'vdd_gpu_s0' already present!
[    4.187463] mali 27800000.gpu: avs=0
[    4.187478] W : [File] : drivers/gpu/arm/bifrost/platform/rk/mali_kbase_config_rk.c; [Line] : 144; [Func] : kbase_platform_rk_init(); power-off-delay-ms not available.
[    4.188104] mali 27800000.gpu: Register LUT 00070200 initialized for GPU arch 0x00070400
[    4.188123] mali 27800000.gpu: GPU identified as 0x2 arch 7.4.0 r1p0 status 0
[    4.188231] mali 27800000.gpu: No priority control manager is configured
[    4.188238] mali 27800000.gpu: Large page allocation set to false after hardware feature check
[    4.188387] mali 27800000.gpu: No memory group manager is configured
[    4.189023] mali 27800000.gpu: EM: OPP:500000 is inefficient
[    4.189029] mali 27800000.gpu: EM: OPP:400000 is inefficient
[    4.189037] mali 27800000.gpu: EM: OPP:300000 is inefficient
[    4.189159] mali 27800000.gpu: EM: created perf domain
[    4.189463] mali 27800000.gpu: l=10000 h=2147483647 hyst=5000 l_limit=0 h_limit=0 h_table=0
[    4.190923] mali 27800000.gpu: * MALI kbase_mmap_min_addr compiled to CONFIG_DEFAULT_MMAP_MIN_ADDR, no runtime update possible! *
[    4.190940] mali 27800000.gpu: Probed as mali0
[    4.306787] I : [File] : drivers/gpu/arm/mali400/mali/linux/mali_kernel_linux.c; [Line] : 407; [Func] : mali_module_init(); svn_rev_string_from_arm of this mali_ko is '', rk_ko_ver is '5', built at '11:51:44', on 'May  6 2024'.
armsom@armsom-sige5:~$ sudo cat /sys/kernel/debug/devices_deferred 
mtd_vendor_storage
```
Currently glmark2 uses llvmpipe, so I will checkout Mesa and try to reverse engineer the exact model for the Mali G52 MC3 variant, I have not been able to find any datasheets for this specific variant.
```
armsom@armsom-sige5:~$ glmark2
** GLX does not support GLX_EXT_swap_control or GLX_MESA_swap_control!
** Failed to set swap interval. Results may be bounded above by refresh rate.
=======================================================
    glmark2 2023.01
=======================================================
    OpenGL Information
    GL_VENDOR:      Mesa/X.org
    GL_RENDERER:    llvmpipe (LLVM 15.0.6, 128 bits)
    GL_VERSION:     4.5 (Compatibility Profile) Mesa 22.3.6
    Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=32 stencil=0 samples=0
    Surface Size:   800x600 windowed
=======================================================
** GLX does not support GLX_EXT_swap_control or GLX_MESA_swap_control!
** Failed to set swap interval. Results may be bounded above by refresh rate.
[build] use-vbo=false: FPS: 150 FrameTime: 6.684 ms
```

**Reply by Joshua-Riek** [reply]
Seems I have forgot to enable CONFIG_MFD_RK806_I2C in the Ubuntu Rockchip kernel config, this will fix the deferred driver and regulator problems.

![image](https://github.com/user-attachments/assets/6582bee2-9832-43ed-8030-7a045596a7b5)

**Comment by as-jackson** [comment]
wow, now Panfrost works.
https://github.com/armbian/linux-rockchip/pull/249
https://github.com/armbian/build/commit/836def3d11699537e6289dedfbc766c2050d422d
